planners/src/solver.rs

```diff
@@ -0,0 +1,269 @@
+use crate::encode::{encode, populate_with_task_network, populate_with_template_instances};
+use crate::fmt::{format_hddl_plan, format_partial_plan, format_pddl_plan};
+use crate::forward_search::ForwardSearcher;
+use crate::solver::Strat::{Activity, Forward};
+use crate::Solver;
+
+use anyhow::{Context, Result};
+use aries_core::state::Domains;
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Instant;
+use structopt::StructOpt;
+
+use aries_model::extensions::SavedAssignment;
+use aries_planning::chronicles::analysis::hierarchical_is_non_recursive;
+use aries_planning::chronicles::Problem;
+use aries_planning::chronicles::*;
+
+use aries_solver::parallel_solver::ParSolver;
+use aries_tnet::theory::{StnConfig, StnTheory, TheoryPropagationLevel};
+
+/// An automated planner for PDDL and HDDL problems.
+#[derive(Debug, Default, Clone, StructOpt)]
+#[structopt(name = "grpc", rename_all = "kebab-case")]
+pub struct Opt {
+    #[structopt(long, short)]
+    pub domain: Option<PathBuf>,
+    /// Path to the problem file.
+    pub problem: PathBuf,
+    /// If set, a machine readable plan will be written to the file.
+    #[structopt(long = "output", short = "o")]
+    plan_out_file: Option<PathBuf>,
+    /// Minimum depth of the instantiation. (depth of HTN tree or number of standalone actions with the same name).
+    #[structopt(long)]
+    min_depth: Option<u32>,
+    /// Maximum depth of instantiation
+    #[structopt(long)]
+    max_depth: Option<u32>,
+    /// If set, the solver will attempt to minimize the makespan of the plan.
+    #[structopt(long = "optimize")]
+    optimize_makespan: bool,
+    /// If true, then the problem will be constructed, a full propagation will be made and the resulting
+    /// partial plan will be displayed.
+    #[structopt(long = "no-search")]
+    no_search: bool,
+    /// If provided, the solver will only run the specified strategy instead of default set of strategies.
+    /// When repeated, several strategies will be run in parallel.
+    #[structopt(long = "strategy", short = "s")]
+    strategies: Vec<Strat>,
+}
+
+pub struct Planner {
+    pub option: Opt,
+    pub problem: Option<FiniteProblem>,
+    pub plan: Option<Arc<Domains>>,
+    pub start: Instant,
+    pub end: Instant,
+    pub htn_mode: bool,
+}
+
+impl Planner {
+    pub fn new(option: Opt) -> Self {
+        Self {
+            option,
+            problem: None,
+            plan: None,
+            start: Instant::now(),
+            end: Instant::now(),
+            htn_mode: false,
+        }
+    }
+
+    pub fn get_answer(&self) -> Option<Arc<Domains>> {
+        self.plan.clone()
+    }
+
+    pub fn solve(&mut self, mut spec: Problem, opt: &Opt) -> Result<()> {
```

Comment:
Make a function: remove the `self` and pass needed parameters explicitly (breaking out the `Opt` into individual parameters). It should not have access to any uneeded information (e.g. the problem filename).

Could be renamed `deepening_solve`.